Implements 'Key concepts' developer documentation for Encrypted Saved Objects (#184334)
elastic/kibana: dev_docs/key_concepts/encrypted_saved_objects.mdx

---
id: kibDevDocsEncryptedSavedObjectsIntro
slug: /kibana-dev-docs/key-concepts/encrypted-saved-objects-intro
title: Encrypted Saved Objects
description: Configure your saved object types to secure sensitive data.
date: 2024-05-21
tags: ['kibana', 'dev', 'contributor', 'api docs']
---

"Encrypted Saved Objects" (ESO) are <DocLink id="kibDevDocsSavedObjectsIntro" text="Saved Object types" /> that have been registered with the Encrypted Saved Objects
Service (ESO Service) to the specify which attributes should be protected (encrypted attributes) and which attributes, if any, should be present and unchanged

> in order to decrypt the protected attributes ("Additional Authenticated Data", or AAD).

Reviewer: QQ: I think for me as the least familiar person with the definition of AAD it is a bit unclear. Does AAD term cover both encrypted attributes and the ones used for decryption or ONLY attributes used for decryption?

internal Kibana server. If end users must have access to the encrypted values, register the encrypted attribute with a `attributesToEncrypt` object and use
`dangerouslyExposeValue: true`.

### AAD

AAD, or Additional Authenticated Data, is part of an "Authenticated Encryption" schema. AAD is an unencrypted string that is used during encryption and decryption
operations in addition to the supplied encryption key, to protect access to encrypted values. If AAD is used during encryption, it must be provided during decryption,
and must be an exact match to the AAD value used during encryption, otherwise decryption will fail. Typically, AAD comprises data that could only be accessed by an
authenticated user and either never changes, or would only potentially change when encrypted values change.

For ESOs, AAD is constructed of key-value pairs composed of the Saved Object Descriptor and any attributes included in `attributesToIncludeInAAD` when the ESO is
registered. The Saved Object Descriptor consists of the object type, ID, and, if applicable, namespace (or space ID). The descriptor for Space-agnostic types

> (`namespaceType: 'agnostic'`) will not include a namespace, and for multi-namespace types (`namespaceType: 'multiple-isolated'` and `namespaceType: 'multiple'`) will

Author: You're right! I saw the handling of a namespace array in getDescriptorNamespace, but I guess this was just to keep the inputs to the function compatible with all possible input.

only include the first namespace found in the object's list of namespaces.

```ts
  export interface SavedObjectDescriptor {
    readonly id: string;
    readonly type: string;
    readonly namespace?: string;
  }
```

#### Nested attributes

When an attribute is included in AAD, all of its properties, or subfields, are inherently included in AAD. When AAD is constructed as key-value pairs, the nested
properties of an attribute are all included in its value. In this way, AAD inclusion is hierarchical.

#### What attributes should be included in AAD?

Determining which attributes to include in AAD is not an exact science, however there are some basic guidelines.

> Good candidates for attributes to INCLUDE in AAD are attributes that...

Author: Made some updates in eb3b16fcd1e79d962530b87ff699fa1cd0d4f97e

  modelVersion: SavedObjectsModelVersion;
  shouldTransformIfDecryptionFails?: boolean;
  inputType: EncryptedSavedObjectTypeRegistration;
  outputType: EncryptedSavedObjectTypeRegistration;
}
```

The `inputType` parameter provides the necessary ESO registration definition for decrypting an ESO of the preceding Model Version prior to performing any transforms
defined by the Model Version. The `outputType` parameter provides the necessary ESO registration definition for encrypting an ESO once the transforms defined in the
Model Version have been completed. All of the Model Version transform functions ('unsafe_transform', 'data_backfill', 'data_removal'), are merged into a single
transform function for efficiency. This way, each ESO only needs to be decrypted and re-encrypted once to incorporate all of the changes defined in a Model Version.
The optional `shouldTransformIfDecryptionFails` parameter defines whether an ESO type should proceed with the Model Version changes even if decryption fails.

Some examples of `createModelVersion` can be found in the ESO Model Version example plugin (examples/eso_model_version_example/server/plugin.ts) [ToDo: better link?]

> For more information regarding Model Versions see the documentation at packages/core/saved-objects/core-saved-objects-server/docs/model_versions.md [ToDo: not sure

Reviewer 2: Looks like we have some docs for model version here: https://docs.elastic.dev/kibana-dev-docs/tutorials/saved-objects#defining-model-versions

this exists in vercel yet?]

#### Serverless Considerations

Changes to ESOs must be carefully considered due to how upgrades occur for Serverless projects. In Serverless, there is a "Zero Downtime" (ZDT) upgrade algorithm. This
means that both the latest and previous versions of Kibana may be running simultaneously. In regard to ESOs, this means that the previous version of Kibana may attempt
to access ESOs that have been migrated by the latest version of Kibana, and in order to do so, must be able to decrypt them successfully without having any knowledge of

> the new Model Version definition or changes to the ESO's `EncryptedSavedObjectTypeRegistration`. Thus, ff an ESO's AAD has changed due to the migration, the previous

Reviewer 2: ```suggestion
the new Model Version definition or changes to the ESO's `EncryptedSavedObjectTypeRegistration`. Thus, if an ESO's AAD has changed due to the migration, the previous
```

version will not be able to decrypt it. It is critical that when changes are made to an ESO, that they either do not affect its AAD or are staged carefully in
subsequent Model Versions.

It is worth noting here that if a ESO's Model Version `forwardCompatibility` schema is set to drop unknown fields (when the `unknowns` option is set to `ignore`),
ESOs of this type will first be decrypted before the unknown fields are dropped. This more easily supports the hierarchical aspect of AAD-included attributes - when
subfields of an attribute are added or removed, the previous version of Kibana will still be able to successfully construct AAD and decrypt the object.

The table below offers some general guidance on how various changes could be supported (or not). Keep in mind that any time you are adding or removing attributes from
an Saved Object type, all related business logic for that type must be capable of gracefully and appropriately handling an object with or without the attribute in both
the current and previous version of Kibana. Some of the advice here is applicable to any Saved Object type migration.

> | Change to ESO                                | Encrypted? | In AAD? | General Guidance                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer 2: This table is awesome ❤️ !

Author: Thanks! I hope it gets used! 🤞

| Add an existing attribute to AAD             | No         | No->Yes | This is not currently supported. Existing attributes that are in use and populated with data cannot be added to AAD. The previous version of Kibana will never be able to successfully perform decryption in this case.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |

> | Remove an existing attribute                 | No         | No      | Implement the Model Version as needed. There is no need to wrap the Model Version with `createModelVersion` because AAD is unaffected. If the previous version of will not be able to tolerate the missing attribute, this will require 2 Serverless release stages. `Release 1`: update all business logic to handle this type without the attribute that will be removed. `Release 2`: implement the Model Version as described.                                                                                                                                                                                                                                                                         |

Reviewer 2: __nit:__ I see in other cases we use `a Model Version`, and here `the Model Version` - is it intentional? 🙂 

> previous version of will 

Previous version of ... __Kibana__? (here and in a few places below)

Author: Some of them were intentional, but not all. I cleaned it up! And yes, I was missing "Kibana" in a few places due to a copy/paste error.

| Modify an attribute (add or remove subfield) | Yes        | N/A     | Implement a Model Version and wrap it in a call to `createModelVersion`. Implement Model Version changes as needed. Even though AAD is unaffected, the objects will require decryption and re-encryption to be modified. If the previous version of will not be able to tolerate the attribute changes, this will require 2 Serverless release stages. `Release 1`: update all business logic to handle this type with the modified attribute. `Release 2`: implement the Model Version as described.                                                                                                                                                                                                      |

> | Change an existing attribute to be encrypted | No->Yes    | No/Yes  | This is not currently supported. The previous version of Kibana will not know to decrypt this attribute, and any business logic that utilizes the attribute will fail.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer 2: ```suggestion
| Change an existing attribute to be encrypted | No->Yes    | No/Yes  | This is not currently supported. The previous version of Kibana will not decrypt this attribute, and any business logic that utilizes the attribute will fail.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
```
or something similar?

<DocLink id="kibDevDocsEncryptedSavedObjectsIntro" section="serverless-considerations" text="Serverless Considerations"/> section, but the basics are:

- An attribute cannot be removed from AAD once it is included, unless it can be altogether removed from the object type, or refactored with a new name.
- An existing attribute cannot be added to AAD if it was not included in AAD when it was first defined and has already been populated.

When making the decision of which attributes to include in AAD, it is best to be conservative and only include attributes that the owning team is 100% confident
should be included.

## Caveats

### Partial Updates

Partial updates on ESOs are only possible if the changes are limited to unencrypted and non-AAD attributes. Any changes to an ESO's encrypted values or AAD-
included values requires re-encryption, which means the entire object must be provided when updating to avoid corrupting the object. If an ESO is corrupted by
a partial update, it will be effectively undecryptable. Currently, there is nothing preventing or limiting partial updates of ESOs (see open GitHub issue

> [50256](https://github.com/elastic/kibana/issues/50256)), so this requires consistent diligence from developers utilizing ESOs.

Reviewer 3: (unrelated to the PR) TIL. I'll reply on the linked issue to reopen the discussion. Things have changed since 2019, we're now doing client-side updates, meaning that we do fetch the attributes from ES for partial updates, so this should be addressable (but maybe it requires encryption to be a first class citizen of the SOR first/\.

input parameters.

```ts
export interface CreateEsoModelVersionFnOpts {
  modelVersion: SavedObjectsModelVersion;
  shouldTransformIfDecryptionFails?: boolean;
  inputType: EncryptedSavedObjectTypeRegistration;
  outputType: EncryptedSavedObjectTypeRegistration;
}
```

The `inputType` parameter provides the necessary ESO registration definition for decrypting an ESO of the preceding Model Version prior to performing any transforms
defined by the Model Version. The `outputType` parameter provides the necessary ESO registration definition for encrypting an ESO once the transforms defined in the
Model Version have been completed. All of the Model Version transform functions ('unsafe_transform', 'data_backfill', 'data_removal'), are merged into a single
transform function for efficiency. This way, each ESO only needs to be decrypted and re-encrypted once to incorporate all of the changes defined in a Model Version.
The optional `shouldTransformIfDecryptionFails` parameter defines whether an ESO type should proceed with the Model Version changes even if decryption fails.

Some examples of `createModelVersion` can be found in the ESO Model Version example plugin (

> [examples/eso_model_version_example/server/plugin.ts](https://github.com/elastic/kibana/blob/06fc22a0f15e692857ba689a7b0ddec91ed2dac2/examples/eso_model_version_example/server/plugin.ts))

Reviewer 3: I would try adding the "simplest" example in this doc, and then link to the example plugin for the other examples

Author: Example added in 94b41de3d540d33a07c15807bb9ec922812ccae5

There were not many options to choose from to include a Model Version implemented ESO, so I just picked the simplest one, even if it was not "perfect" (we don't really have any perfectly ideal ESO definitions).

| Remove an attribute from AAD inclusion       | No         | Yes->No | This is not currently supported. The previous version of Kibana will always use the attribute to construct AAD.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

## A real world example

Let's examine one of the latest ESO types as of writing this document, the 'ad_hoc_run_params' type. We're using this type as an example, because it is one of the first
new ESO types to use a Model Version, and it is a fairly simple type (compared to other Model Version ESOs).

### Registering the Saved Object type

> Below is the call to reigster the type.

Reviewer: `reigster` -> `register`

> Below is the call to reigster the type.

Reviewer: ```suggestion
Below is the call to register the type.
```

    spaceId: schema.string(),
    start: schema.string(),
    status: rawAdHocRunStatus,
    schedule: schema.arrayOf(rawAdHocRunSchedule),
  });

  const rawAdHocRunParamsRuleSchema = schema.object({
    name: schema.string(),
    tags: schema.arrayOf(schema.string()),
    alertTypeId: schema.string(),
    params: schema.recordOf(schema.string(), schema.maybe(schema.any())),
    apiKeyOwner: schema.nullable(schema.string()),
    apiKeyCreatedByUser: schema.maybe(schema.nullable(schema.boolean())),
    consumer: schema.string(),
    enabled: schema.boolean(),
    schedule: schema.object({
      interval: schema.string(),
    }),
    createdBy: schema.nullable(schema.string()),
    updatedBy: schema.nullable(schema.string()),
    updatedAt: schema.string(),
    createdAt: schema.string(),
    revision: schema.number(),
  });

  const rawAdHocRunStatus = schema.oneOf([
    schema.literal('complete'),
    schema.literal('pending'),
    schema.literal('running'),
    schema.literal('error'),
    schema.literal('timeout'),
  ]);

  const rawAdHocRunSchedule = schema.object({
    interval: schema.string(),
    status: rawAdHocRunStatus,
    runAt: schema.string(),
  });
```

> ### Registering the type as an Encrytped Saved Object

Reviewer: `Encrytped` -> `Encrypted`

> ### Registering the type as an Encrytped Saved Object

Reviewer: ```suggestion
### Registering the type as an Encrypted Saved Object
```

```ts
  encryptedSavedObjects.registerType({
    type: AD_HOC_RUN_SAVED_OBJECT_TYPE, // 'ad_hoc_run_params' - the type name must match
    attributesToEncrypt: new Set(['apiKeyToUse']),
    attributesToIncludeInAAD: new Set(['rule', 'spaceId']),
  });
```

There is only one encrypted attribute, `apiKeyToUse`, which contains the API key that needs to be protected. The attributes to include in AAD consist of the `spaceId`,
and the `rule` definition. Since the object's namespace type is `'multiple-isolated'`, the space ID (or namespace) will not automatically be part of AAD by way of the
Saved Object Descriptor (see <DocLink id="kibDevDocsEncryptedSavedObjectsIntro" section="aad" text="AAD"/>). In theory, the space id will never change for a

> `'multiple-isolated'`, which makes it a reasonable candidate for an AAD field.

Author: I also debated including some explanation regarding whether adding `spaceId` to AAD is offering any real protection in this case.

"...but it would also require some additional management to ensure it contributes to protecting the data. For example, if an object is moved to a different space maliciously (or accidentally) in order to try to gain access to its encrypted attributes, the `spaceId` attribute does not necessarily reflect the space the object resides in, and on its own provides no mechanism to guarantee the object has not moved. Some additional mechanism to override the attribute's value with the object's current namespace prior to attempting decryption would add protection."

But I don't know if this is just too much, or if what I am saying here is even feasible to implement.

cc @azasypkin @pgayvallet

Author: I am going to merge the doc as-is, and potentially address this in an update.

> The `rule` attribute contains values related to the encrtypted attribute `apiKeyToUse`, or that will never change:

Reviewer: ```suggestion
The `rule` attribute contains values related to the encrypted attribute `apiKeyToUse`, or that will never change:
```

```ts
  apiKeyOwner: schema.nullable(schema.string()),
  apiKeyCreatedByUser: schema.maybe(schema.nullable(schema.boolean())),
  createdBy: schema.nullable(schema.string()),
  createdAt: schema.string(),
```

The above are prime examples of the type of attributes that should comprise AAD. The `rule` attribute also contains some attributes that may change independently of the
encrypted attribute `apiKeyToUse`:

```ts
  enabled: schema.boolean(),
  schedule: schema.object({
    interval: schema.string(),
  }),
  updatedBy: schema.nullable(schema.string()),
  updatedAt: schema.string(),
  revision: schema.number(),
```

This is not a problem, but it is important to consider that a change to any of these attributes will require re-encryption of an object. It is worth considering the nature

> of AAD hierarchical inclusion when structuring attributes for your saved objects. You can also utilize more granular keys when speficying which attributes to include in AAD,

Reviewer: ```suggestion
of AAD hierarchical inclusion when structuring attributes for your saved objects. You can also utilize more granular keys when specifying which attributes to include in AAD,
```

e.g. `rule.apiKeyOwner`. For more information, see the <DocLink id="kibDevDocsEncryptedSavedObjectsIntro" section="nested-attributes" text="Nested attributes"/> section of
this document.

> Additionally, the owning team implemeted a type to help manage partial updates. This is a great addition to ensure changes to the ESOs do not render them un-decryptable.

Reviewer: ```suggestion
Additionally, the owning team implemented a type to help manage partial updates. This is a great addition to ensure changes to the ESOs do not render them un-decryptable.
```

> Additionally, the owning team implemeted a type to help manage partial updates. This is a great addition to ensure changes to the ESOs do not render them un-decryptable.

Reviewer: ```suggestion
Additionally, the owning team implemeted a type to help manage partial updates. This is a great addition to ensure changes to the ESOs do not render them undecryptable.
```

```ts
export type AdHocRunAttributesNotPartiallyUpdatable = 'rule' | 'spaceId' | 'apiKeyToUse';
```

Usage:
```ts

> export type PartiallyUpdateableAdHocRunAttributes = Partial<

Author: This one is straight from the repo. I think it is correct. Thanks for proof reading! ❤️
